flatgfa/src/cmds.rs

```diff
@@ -318,3 +320,199 @@ pub fn depth(gfa: &flatgfa::FlatGFA) {
+                    overlap
+                }
+            })
+        );
+    }
+
+    // Add new, chopped segments
+    for seg in gfa.segs.all().iter() {
+        let len = seg.len();
+        if len <= args.c {
+            // Leave the segment as is
+            let id = flat.segs.add(Segment {
+                name: max_node_id,
+                seq: seg.seq,
+                optional: empty_span()
+                // TODO: Optional data may stay valid when seg not chopped?
+            });
+            max_node_id += 1;
+            seg_map.push((id, flat.segs.next_id()));
+        }
+        else {
+            let seq_end = seg.seq.end;
+            let mut offset = seg.seq.start.index();
+            let segs_start = flat.segs.next_id();
+            // Could also generate end_id by setting it equal to the start_id and
+            // updating it for each segment that is added - only benefits us if we 
+            // don't unroll the last iteration of this loop
+            while offset < seq_end.index() - args.c {
+                // Generate a new segment of length c
+                flat.segs.add(Segment {
+                    name: max_node_id,
+                    seq: Span::new(Id::new(offset), Id::new(offset + args.c)),
+                    optional: empty_span()
+                });
+                offset += args.c;
+                max_node_id += 1;
+            }
+            // Generate the last segment
+            flat.segs.add(Segment {
+                    name: max_node_id,
+                    seq: Span::new(Id::new(offset), seq_end),
+                    optional: empty_span()
+            });
+            max_node_id += 1;
+            let new_seg_range = (segs_start, flat.segs.next_id());
+            seg_map.push(new_seg_range);
+            if args.l {
+                link_forward(&mut flat, &new_seg_range);
+            }
+        }
+    }
+
+    // For each path, add updated handles. Then add the updated path
+    for path in gfa.paths.all().iter() {
+        let path_start = flat.steps.next_id();
+        let mut path_end = flat.steps.next_id();
+        // Generate the new handles
+        // Tentative to-do: see if it is faster to read Id from segs than to re-generate it?
+        for step in &gfa.steps[path.steps] {
+            let (start_id, end_id) = seg_map[step.segment().index()];
+            let (start_idx, end_idx) = (start_id.index(), end_id.index());
+            match step.orient() {
+                Orientation::Forward => {
+                    // In this builder, Id.index() == seg.name - 1 for all seg
+                    path_end = flat.add_steps(
+                        (start_idx..end_idx).map(|idx| {
+                            Handle::new(
+                                Id::new(idx),
+                                Orientation::Forward
+                            )
+                        })
+                    ).end;
```

Comment:
DEFINITELY not for this PR, but: I wonder if there's some kind of utility method that we can invent to help with this stuff. Impressionistically speaking, what we want here is...

```rs
let segs = seg_map[step.segment()];
flat.add_steps(segs.map(|s| Handle::new(s, Orientation::Forward)));
```

Like, we kind of want a way to do a `map` directly on a chunk of segments, without having to fiddle with the index math here. Maybe we can think of a clever way to make that look nice!

Reply:
Makes sense! `seg_map` currently maps from indexes (or segment names) to a tuple of `Id`s, which preserves some type safety but could be better (if I wrap `Segment.name` in an `Id`, it would provide more type safety but I think I'd have to use a HashMap instead of a vector. Maybe there's a trait that could be implemented to automatically convert `Id`s to `ints`? Also, I had been thinking of the `Id` wrapper as a zero cost abstraction, but actually, wouldn't it need to be stored on the heap instead of the stack, and wouldn't this add overhead in some cases?).

Anyways, this probably means that doing a `map` directly on a chunk of segments is not particularly helpful here, since it would still entail mapping from indexes to segments. However, I think that implementing `Range` or something similar, such that we can write something like `start_id..end_id.map(|id| { Handle::new(Id, Orientation::Forward })` would help make this function more readable. In general, it seems like being able to treat `Id`s like regular numbers is a good thing (as long as they're distinct to the type checker). Something like this could probably be implemented in a small follow-up PR?

Reply:
Maybe this is silly, but could the RHS of `seg_map` contain `Span<Segment>` instead of `(Id<Segment>, Id<Segment>)`? Then we could perhaps implement a `map` function on `Span`s...